File: worker-python/app/worker.py
import json
import time
import argparse
import mimetypes
import subprocess
import sys
import os
from pathlib import Path

import pandas as pd


def emit(event_type, message=None, value=None, extra=None):
    payload = {"type": event_type}
    if message is not None:
        payload["message"] = message
    if value is not None:
        payload["value"] = value
    if extra is not None:
        payload["extra"] = extra
    print(json.dumps(payload), flush=True)


def fail(message, code=1):
    emit("error", message=message)
    raise SystemExit(code)


def list_artifacts(artifacts_dir: Path):
    items = []
    if not artifacts_dir.exists():
        return items

    for p in sorted(artifacts_dir.rglob("*")):
        if p.is_file():
            mime, _ = mimetypes.guess_type(str(p))
            items.append({
                "name": p.name,
                "path": str(p),
                "relative_path": str(p.relative_to(artifacts_dir)),
                "type": mime or "application/octet-stream",
                "size_bytes": p.stat().st_size
            })
    return items


def load_dataframe(data_format, data_path):
    fmt = data_format.lower().strip()
    path = Path(data_path)

    if not path.exists():
        fail(f"Arquivo não encontrado: {data_path}")

    emit("log", message=f"Lendo arquivo no formato {fmt}: {data_path}")
    emit("progress", value=15, message="Iniciando leitura da base")

    if fmt == "csv":
        df = pd.read_csv(path)
    elif fmt == "parquet":
        df = pd.read_parquet(path)
    elif fmt == "json":
        df = pd.read_json(path)
    elif fmt in ("xlsx", "excel"):
        df = pd.read_excel(path)
    else:
        fail(f"Formato não suportado neste MVP: {data_format}")

    emit("progress", value=55, message="Base carregada em memória")
    return df


def write_json(path: Path, data: dict):
    with open(path, "w", encoding="utf-8") as f:
        json.dump(data, f, ensure_ascii=False, indent=2)


def execute_user_code(code_file: Path, output_dir: Path):
    if not code_file.exists():
        fail(f"Código não encontrado: {code_file}")

    artifacts_dir = output_dir / "artifacts"
    artifacts_dir.mkdir(parents=True, exist_ok=True)

    emit("log", message=f"Executando código do usuário: {code_file}")
    emit("progress", value=20, message="Iniciando execução do código")

    env = os.environ.copy()
    env["OUTPUT_DIR"] = str(output_dir)
    env["ARTIFACTS_DIR"] = str(artifacts_dir)

    proc = subprocess.run(
        [sys.executable, str(code_file)],
        cwd=str(code_file.parent),
        env=env,
        capture_output=True,
        text=True
    )

    (output_dir / "stdout.log").write_text(proc.stdout, encoding="utf-8")
    (output_dir / "stderr.log").write_text(proc.stderr, encoding="utf-8")

    artifacts = list_artifacts(artifacts_dir)

    result = {
        "status": "success" if proc.returncode == 0 else "failed",
        "mode": "inline_code_execution",
        "returncode": proc.returncode,
        "stdout_log": str(output_dir / "stdout.log"),
        "stderr_log": str(output_dir / "stderr.log"),
        "artifacts": artifacts,
        "stdout_preview": proc.stdout[:2000],
        "stderr_preview": proc.stderr[:2000]
    }

    write_json(output_dir / "result.json", result)

    emit("log", message=f"stdout salvo em {output_dir / 'stdout.log'}")
    emit("log", message=f"stderr salvo em {output_dir / 'stderr.log'}")
    emit("log", message=f"Artefatos encontrados: {len(artifacts)}")
    emit("progress", value=100, message="Execução do código concluída")
    emit("result", message="Execução finalizada", extra={"result_file": str(output_dir / 'result.json')})

    if proc.returncode != 0:
        fail("Código do usuário terminou com erro", proc.returncode)


def main():
    parser = argparse.ArgumentParser(description="Worker analítico Python")
    parser.add_argument("--data-format", required=False, default=None)
    parser.add_argument("--data-path", required=False, default=None)
    parser.add_argument("--ram-limit-gb", required=False, default="2")
    parser.add_argument("--cpus", required=False, default="1")
    parser.add_argument("--output-dir", required=False, default="/output")
    parser.add_argument("--code-file", required=False, default=None)
    args = parser.parse_args()

    output_dir = Path(args.output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    emit("log", message="Worker Python iniciado")
    emit(
        "log",
        message="Parâmetros recebidos",
        extra={
            "data_format": args.data_format,
            "data_path": args.data_path,
            "ram_limit_gb": args.ram_limit_gb,
            "cpus": args.cpus,
            "output_dir": str(output_dir),
            "code_file": args.code_file
        },
    )

    time.sleep(1)

    if args.code_file:
        execute_user_code(Path(args.code_file), output_dir)
        return

    if args.data_path and args.data_format:
        df = load_dataframe(args.data_format, args.data_path)

        emit("progress", value=75, message="Processando estatísticas")
        time.sleep(1)

        result = {
            "status": "success",
            "mode": "with_data_source",
            "rows": int(df.shape[0]),
            "columns": int(df.shape[1]),
            "column_names": list(df.columns),
            "memory_usage_bytes": int(df.memory_usage(deep=True).sum()),
            "preview": df.head(10).to_dict(orient="records"),
            "artifacts": []
        }
    else:
        emit("log", message="Execução sem base de dados")
        emit("progress", value=50, message="Executando rotina sem data source")
        time.sleep(1)

        result = {
            "status": "success",
            "mode": "no_data_source",
            "message": "Execução concluída sem carregar base",
            "rows": 0,
            "columns": 0,
            "column_names": [],
            "memory_usage_bytes": 0,
            "preview": [],
            "artifacts": []
        }

    write_json(output_dir / "result.json", result)

    emit("log", message=f"Resultado salvo em {output_dir / 'result.json'}")
    emit("progress", value=100, message="Processamento concluído")
    emit("result", message="Execução finalizada", extra={"result_file": str(output_dir / 'result.json')})


if __name__ == "__main__":
    main()

File: worker-r/app/worker.R
#!/usr/bin/env Rscript

args <- commandArgs(trailingOnly = TRUE)

get_arg_value <- function(flag, default = NULL) {
  idx <- match(flag, args)
  if (!is.na(idx) && idx < length(args)) {
    return(args[idx + 1])
  }
  return(default)
}

emit <- function(event_type, message = NULL, value = NULL, extra = NULL) {
  payload <- list(type = event_type)
  if (!is.null(message)) payload$message <- message
  if (!is.null(value)) payload$value <- value
  if (!is.null(extra)) payload$extra <- extra

  cat(jsonlite::toJSON(payload, auto_unbox = TRUE), "\n")
  flush.console()
}

fail <- function(message, code = 1) {
  emit("error", message = message)
  quit(status = code)
}

list_artifacts <- function(artifacts_dir) {
  if (!dir.exists(artifacts_dir)) {
    return(list())
  }

  files <- list.files(artifacts_dir, recursive = TRUE, full.names = TRUE)
  files <- files[file.info(files)$isdir == FALSE]

  result <- list()
  for (f in files) {
    rel <- sub(
      paste0("^", normalizePath(artifacts_dir, winslash = "/"), "/?"),
      "",
      normalizePath(f, winslash = "/")
    )
    size <- file.info(f)$size

    result[[length(result) + 1]] <- list(
      name = basename(f),
      path = f,
      relative_path = rel,
      type = if (grepl("\\.png$", f, ignore.case = TRUE)) "image/png"
             else if (grepl("\\.jpg$|\\.jpeg$", f, ignore.case = TRUE)) "image/jpeg"
             else if (grepl("\\.svg$", f, ignore.case = TRUE)) "image/svg+xml"
             else if (grepl("\\.json$", f, ignore.case = TRUE)) "application/json"
             else "application/octet-stream",
      size_bytes = unname(size)
    )
  }

  return(result)
}

write_json <- function(path, data) {
  jsonlite::write_json(data, path = path, auto_unbox = TRUE, pretty = TRUE)
}

execute_user_code <- function(code_file, output_dir) {
  if (!file.exists(code_file)) {
    fail(paste("Código não encontrado:", code_file))
  }

  artifacts_dir <- file.path(output_dir, "artifacts")
  dir.create(artifacts_dir, recursive = TRUE, showWarnings = FALSE)

  emit("log", message = paste("Executando código do usuário:", code_file))
  emit("progress", message = "Iniciando execução do código", value = 20)

  stdout_file <- file.path(output_dir, "stdout.log")
  stderr_file <- file.path(output_dir, "stderr.log")

  result_code <- system2(
    command = "Rscript",
    args = c(code_file),
    stdout = stdout_file,
    stderr = stderr_file,
    env = c(
      paste0("OUTPUT_DIR=", output_dir),
      paste0("ARTIFACTS_DIR=", artifacts_dir)
    )
  )

  artifacts <- list_artifacts(artifacts_dir)

  stdout_preview <- ""
  stderr_preview <- ""

  if (file.exists(stdout_file)) {
    stdout_preview <- paste(readLines(stdout_file, warn = FALSE), collapse = "\n")
  }

  if (file.exists(stderr_file)) {
    stderr_preview <- paste(readLines(stderr_file, warn = FALSE), collapse = "\n")
  }

  result <- list(
    status = if (result_code == 0) "success" else "failed",
    mode = "inline_code_execution_r",
    returncode = result_code,
    stdout_log = stdout_file,
    stderr_log = stderr_file,
    artifacts = artifacts,
    stdout_preview = substr(stdout_preview, 1, 2000),
    stderr_preview = substr(stderr_preview, 1, 2000)
  )

  write_json(file.path(output_dir, "result.json"), result)

  emit("log", message = paste("stdout salvo em", stdout_file))
  emit("log", message = paste("stderr salvo em", stderr_file))
  emit("log", message = paste("Artefatos encontrados:", length(artifacts)))
  emit("progress", message = "Execução do código concluída", value = 100)
  emit("result", message = "Execução finalizada", extra = list(result_file = file.path(output_dir, "result.json")))

  if (result_code != 0) {
    fail("Código do usuário em R terminou com erro", result_code)
  }
}

main <- function() {
  data_format <- get_arg_value("--data-format", NULL)
  data_path <- get_arg_value("--data-path", NULL)
  ram_limit_gb <- get_arg_value("--ram-limit-gb", "2")
  cpus <- get_arg_value("--cpus", "1")
  output_dir <- get_arg_value("--output-dir", "/output")
  code_file <- get_arg_value("--code-file", NULL)

  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)

  emit("log", message = "Worker R iniciado")
  emit(
    "log",
    message = "Parâmetros recebidos",
    extra = list(
      data_format = data_format,
      data_path = data_path,
      ram_limit_gb = ram_limit_gb,
      cpus = cpus,
      output_dir = output_dir,
      code_file = code_file
    )
  )

  Sys.sleep(1)

  if (!is.null(code_file) && nzchar(code_file)) {
    execute_user_code(code_file, output_dir)
    return(invisible(NULL))
  }

  emit("log", message = "Execução sem código R informado")
  emit("progress", message = "Finalizando sem código", value = 100)

  result <- list(
    status = "success",
    mode = "no_code_r",
    message = "Execução concluída sem código R",
    artifacts = list()
  )

  write_json(file.path(output_dir, "result.json"), result)
  emit("result", message = "Execução finalizada", extra = list(result_file = file.path(output_dir, "result.json")))
}

suppressPackageStartupMessages({
  if (!requireNamespace("jsonlite", quietly = TRUE)) {
    quit(status = 2)
  }
})

main()
